### CHI Squared Test

In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
import scipy.stats as stats

In [4]:
import seaborn as sns

In [5]:
data = sns.load_dataset('tips')

In [6]:
data.head()

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4



checking whether there is a relation between sex and smoker :- 

In [8]:
dataset_table = pd.crosstab(data["sex"],data["smoker"])

In [9]:
dataset_table

smoker  Yes  No
sex            
Male     60  97
Female   33  54

In [10]:
dataset_table.values

array([[60, 97],
       [33, 54]], dtype=int64)

In [11]:
Observed_values = dataset_table.values

In [12]:
val = stats.chi2_contingency(dataset_table)

In [13]:
val

(0.008763290531773594, 0.925417020494423, 1, array([[59.84016393, 97.15983607],
        [33.15983607, 53.84016393]]))

In [21]:
expected_value = val[3]

In [15]:
no_of_rows = len(dataset_table.iloc[0:2,0])

In [16]:
no_of_columns = len(dataset_table.iloc[0,0:2])

In [17]:
degrees_of_freedom = (no_of_rows-1)*(no_of_columns-1)

In [18]:
degrees_of_freedom

1

In [19]:
alpha= 0.05

chi squared test formula = 

    X**2 = sum((observed value - Expected value)**2/(Expected value))

In [20]:
from scipy.stats import chi2

In [22]:
chi_square = sum([(o-e)**2./e for o,e in zip(Observed_values,expected_value)])

In [23]:
chi_square

array([0.00119737, 0.00073745])

In [24]:
chi_square_statistics = chi_square[0] + chi_square[1]

In [25]:
chi_square_statistics

0.001934818536627623

In [26]:
critical_value = chi2.ppf(q=1-alpha,df=degrees_of_freedom)

In [27]:
critical_value

3.841458820694124

In [28]:
p_value = 1-chi2.cdf(x=chi_square_statistics,df=degrees_of_freedom)

In [29]:
p_value

0.964915107315732

p_value = 0.964915107315732


significance level = alpha = 0.05


degrees_of_freedom = 1


critical_value = 3.841458820694124

In [33]:
if chi_square_statistics >= critical_value:
    print("Reject H0, there is a relation between two categorical varibles")
else:
    print("Retain H0, there is no relation between two categorical variables ")
    


Retain H0, there is no relation between two categorical variables 


In [34]:
if p_value<= alpha:
    print("Reject H0, there is a relation between categorivcal variables")
else :
    print("Retain H0, there is no relation between two categorical values ")

Retain H0, there is no relation between two categorical values 
